From fe2f8ec7fcc0ff3946685fd27cd4c1c5502ed00c Mon Sep 17 00:00:00 2001
From: Talon396 <105325988+Talon396@users.noreply.github.com>
Date: Wed, 27 Jul 2022 07:01:27 -0500
Subject: [PATCH] owlOS Port

Signed-off-by: Talon396 <105325988+Talon396@users.noreply.github.com>
---
 configure | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/configure b/configure
index 421cf859..7a1669fa 100755
--- a/configure
+++ b/configure
@@ -6423,6 +6423,10 @@ CF_EOF
 		LINK_PROGS="$SHELL ${rel_builddir}/mk_prog.sh"
 		LINK_TESTS="$SHELL ${rel_builddir}/mk_prog.sh"
 		;;
+	(owlos*)
+	    CC_SHARED_OPTS='-fPIC'
+	    MK_SHARED_LIB='${CC} -shared -o $@'
+	    ;;
 	(mingw*)
 		cf_cv_shlib_version=mingw
 		cf_cv_shlib_version_infix=mingw
-- 
2.34.1

